# Анализ рынка видеоигр (2000-2013)

### Цели и задачи проекта

Цель проекта - изучить развитие игровой индустрии в период с 2000 по 2013 год на основе исторических данных о продажах видеоигр.

Для достижения этой цели необходимо решить следующие задачи:

- загрузить и изучить исходные данные;
- проверить корректность данных и выполнить их предобработку;
- обработать пропущенные значения и исправить возможные ошибки;
- привести названия столбцов и типы данных к удобному виду;
- выявить и устранить дубликаты;
- отфильтровать данные за период 2000–2013 годов;
- провести категоризацию игр по оценкам пользователей и критиков;
- определить топ-7 игровых платформ по количеству выпущенных игр.

Результаты исследования позволяют оценить качество исходных данных, особенности развития рынка видеоигр и подготовить аналитический срез для дальнейшего анализа. 

### Используемые технологии

- Python
- Pandas
- Jupyter Notebook
- Исследовательский анализ данных (EDA)
- Обработка пропущенных значений
- Работа с дубликатами
- Категоризация данных

### Тип проекта

- Предобработка данных и исследовательский анализ

### Описание данных

Датасет содержит информацию о видеоиграх, выпущенных на различных игровых платформах.

Описание столбцов:

- **name** - название игры
- **platform** - игровая платформа
- **year_of_release** - год выпуска игры
- **genre** - жанр игры
- **na_sales** - продажи в Северной Америке (миллионы копий)
- **eu_sales** - продажи в Европе (миллионы копий)
- **jp_sales** - продажи в Японии (миллионы копий)
- **other_sales** - продажи в других регионах
- **critic_score** - оценка критиков (0–100)
- **user_score** - оценка пользователей (0–10)
- **rating** - возрастной рейтинг ESRB

### План исследования

1. Загрузка и первичное изучение данных.
2. Проверка качества данных.
3. Предобработка:
   - исправление типов данных;
   - обработка пропусков;
   - поиск дубликатов;
   - нормализация значений.
4. Формирование актуального среза данных (2000–2013).
5. Категоризация игр по оценкам пользователей и критиков.
6. Определение наиболее популярных игровых платформ.
7. Формирование выводов.

## Загрузка данных и первичное изучение


In [1]:
# импортируем библиотеку pandas
import pandas as pd

# загружаем датасет
df = pd.read_csv('../data/new_games.csv')

# выводим первые строки таблицы
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [2]:
# выводим информацию о датасете
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  str    
 1   Platform         16956 non-null  str    
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  str    
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  str    
 6   JP sales         16956 non-null  str    
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  str    
 10  Rating           10085 non-null  str    
dtypes: float64(4), str(7)
memory usage: 1.4 MB


### Вывод по первичному изучению данных

Датасет содержит 16956 строк и 11 столбцов.

В таблице представлены данные о видеоиграх: название, платформа, год выпуска, жанр, продажи в различных регионах, а также оценки пользователей и критиков и возрастной рейтинг.

При первичном изучении данных можно заметить несколько особенностей:

- в некоторых столбцах присутствуют пропущенные значения (например, Year of Release, Critic Score, User Score, Rating);
- столбцы EU sales и JP sales имеют тип object, хотя должны содержать числовые значения;
- столбец User Score также имеет тип object, что может быть связано с наличием строковых значений;
- названия столбцов записаны в разных стилях и содержат пробелы, что неудобно для работы.

---

## Проверка ошибок в данных и их предобработка


### Проверка названий столбцов

На данном этапе проверяется корректность названий столбцов и выполняется приведение их к единому стилю snake_case.

In [3]:
# Выводим на экран названия всех столбцов
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='str')

In [4]:
# Приводим столбцы к стилю snake case, регистр - нижний, вместо пробелов - подчеркивания
df.columns = df.columns.str.lower().str.replace(' ', '_')

df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='str')

### Типы данных

In [5]:
# Выводим информацию о датасете для анализа типов данных
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  str    
 1   platform         16956 non-null  str    
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  str    
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  str    
 6   jp_sales         16956 non-null  str    
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  str    
 10  rating           10085 non-null  str    
dtypes: float64(4), str(7)
memory usage: 1.4 MB


### Вывод по результатам анализа типов данных

После изучения типов данных с помощью метода `info()` можно заметить, что некоторые столбцы имеют некорректные типы данных.

Например:

- столбцы `eu_sales` и `jp_sales` имеют тип object, хотя содержат числовые значения;
- столбец `user_score` также имеет тип object, хотя должен содержать числовые оценки пользователей.

Причиной появления типа `object` в числовых столбцах может быть наличие строковых значений в данных. Например, в столбце `user_score` могут встречаться текстовые значения, которые мешают корректному определению типа данных.

Для дальнейшего анализа необходимо привести эти столбцы к числовому типу данных.

In [6]:
# Приводим столбец "Оценка пользователей" к числовому типу и заменяем строковые значения на пропуски
df['user_score'] = pd.to_numeric(df['user_score'], errors='coerce')

In [7]:
# Приводим столбцы с числом продаж к числовому типу и заменяем строковые значения на пропуски
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors='coerce')
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors='coerce')

In [8]:
# Выводим информацию о датасете для проверки результатов
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  str    
 1   platform         16956 non-null  str    
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  str    
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  str    
dtypes: float64(7), str(4)
memory usage: 1.4 MB


### Вывод по результатам анализа типов данных

В ходе проверки типов данных было выявлено несколько несоответствий.

Некоторые столбцы имели тип `object`, хотя должны содержать числовые значения. Вероятнее всего, из-за наличия строковых значений.

Столбцы `user_score`, `eu_sales` и `jp_sales` были преобразованы в числовой тип данных с помощью функции `pd.to_numeric()`. При преобразовании строковые значения были заменены на пропуски при помощи `errors='coerce'`.

После выполнения преобразований типы данных стали корректными, что позволит использовать их для дальнейшего анализа.

### Наличие пропусков в данных

In [9]:
# Считаем количество пропусков (абсолютные значения)
df.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [10]:
# Считаем количество пропусков (относительные значения)
(df.isna().sum() / len(df)) * 100

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

### Комметарий по результатам анализа пропущенных значений

В данных присутствуют пропуски в нескольких столбцах:

- year_of_release
- critic_score
- user_score
- rating
- name
- genre

Причины появления пропусков могут быть следующими:

- для некоторых старых игр отсутствуют оценки критиков и пользователей;
- рейтинг ESRB мог не присваиваться;
- информация о годе выпуска могла быть утеряна при сборе данных.

Строки с пропущенными значениями в столбцах name и genre не несут аналитической ценности, поэтому их можно удалить.

Пропуски в оценках критиков и пользователей можно оставить, так как они не критичны для дальнейшего анализа.

In [11]:
# Удаляем строки с пропусками в названии и жанре
df = df.dropna(subset=['name', 'genre'])

In [12]:
# Повторно считаем пропуски для проверки результатов
df.isna().sum()

name                  0
platform              0
year_of_release     275
genre                 0
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8712
user_score         9266
rating             6869
dtype: int64

### Явные и неявные дубликаты в данных

### Проверка неявных дубликатов

In [13]:
df['genre'].unique()

<StringArray>
[      'Sports',     'Platform',       'Racing', 'Role-Playing',
       'Puzzle',         'Misc',      'Shooter',   'Simulation',
       'Action',     'Fighting',    'Adventure',     'Strategy',
         'MISC', 'ROLE-PLAYING',       'RACING',       'ACTION',
      'SHOOTER',     'FIGHTING',       'SPORTS',     'PLATFORM',
    'ADVENTURE',   'SIMULATION',       'PUZZLE',     'STRATEGY']
Length: 24, dtype: str

In [14]:
df['platform'].unique()

<StringArray>
[ 'Wii',  'NES',   'GB',   'DS', 'X360',  'PS3',  'PS2', 'SNES',  'GBA',
  'PS4',  '3DS',  'N64',   'PS',   'XB',   'PC', '2600',  'PSP', 'XOne',
 'WiiU',   'GC',  'GEN',   'DC',  'PSV',  'SAT',  'SCD',   'WS',   'NG',
 'TG16',  '3DO',   'GG', 'PCFX']
Length: 31, dtype: str

In [15]:
df['rating'].unique()

<StringArray>
['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP']
Length: 9, dtype: str

### Вывод по результатам анализа неявных дубликатов

Для поиска неявных дубликатов были изучены уникальные значения в категориальных столбцах `genre`, `platform` и `rating`.

В столбце `genre` обнаружены значения, которые отличаются только регистром букв. Например:
- `Sports` и `SPORTS`
- `Action` и `ACTION`
- `Role-Playing` и `ROLE-PLAYING`

Такие значения являются неявными дубликатами, так как обозначают один и тот же жанр, но записаны по-разному.

В столбце `platform` дубликатов, связанных с разным написанием, не обнаружено. Названия платформ представлены корректно.

В столбце `rating` также не обнаружено неявных дубликатов, однако присутствуют пропущенные значения.

### Нормализация данных <a class="tocSkip">

In [16]:
# Приводим текст в столбцах к одному регистру
df['genre'] = df['genre'].str.lower()
df['name'] = df['name'].str.lower()
df['rating'] = df['rating'].str.upper()

In [17]:
# Проверка результатов нормализации данных
df['genre'].unique()

<StringArray>
[      'sports',     'platform',       'racing', 'role-playing',
       'puzzle',         'misc',      'shooter',   'simulation',
       'action',     'fighting',    'adventure',     'strategy']
Length: 12, dtype: str

### Итог нормализации неявных дубликатов 
После нормализации текстовых значений жанры игр были приведены к нижнему регистру. 
Это позволило устранить неявные дубликаты, связанные с различиями в регистре букв.

Названия игр также были приведены к нижнему регистру для единообразия данных.

Значения рейтинга ESRB были приведены к верхнему регистру, так как именно в таком виде они обычно используются.

### Проверка явных дубликатов

In [18]:
# Выводим явные дубликаты
df.duplicated().sum()

np.int64(241)

In [19]:
# Удаляем явные дубликаты
df = df.drop_duplicates()

In [20]:
# Результаты проверки удаления явных дубликатов
df.duplicated().sum()

np.int64(0)

### Промежуточный вывод по дубликатам 

В ходе анализа данных были обнаружены неявные дубликаты в столбце `genre`, связанные с различием регистра букв. 
Для устранения этой проблемы текстовые значения были нормализованы: жанры и названия игр приведены к нижнему регистру, а рейтинг ESRB — к верхнему.

После нормализации была проведена проверка на наличие явных дубликатов строк. 
Был обнаружен 241 дубликат.

Найденные дубликаты были удалены из датафрейма с помощью метода `drop_duplicates()`.

После удаления повторная проверка показала, что явные дубликаты отсутствуют.

### Анализ удаленных строк <a class="tocSkip">

In [21]:
# Изначальное количество строк в датасете 
initial_rows = 16956

In [22]:
# Количество строк в датасете после удаления
len(df)

16713

In [23]:
# Количество удаленных строк (абсолютное значение)
absolute_value = initial_rows - len(df) 
absolute_value

243

In [24]:
# Количество удаленных строк (относительное значение)
removed_rows = 243
relative_value = round(removed_rows / initial_rows * 100, 2)
relative_value

1.43

### Общий промежуточный вывод по предобработке данных 

В ходе предобработки данных были выполнены следующие действия:

- названия столбцов были приведены к стилю snake_case;
- исправлены типы данных в числовых столбцах;
- строковые значения в числовых столбцах были заменены на пропуски;
- были изучены пропущенные значения и приняты решения по их обработке;
- проведена нормализация текстовых данных для устранения неявных дубликатов;
- выявлены и удалены явные дубликаты строк.

Всего было удалено 243 строки, что составляет около 1.43% от общего объёма данных.

После выполнения всех этапов данные стали более корректными и пригодными для дальнейшего анализа.

---

## Фильтрация данных

In [25]:
# Формируем отдельный датафрейм с периодом выхода игр с 2000 по 2013 года (включительно)
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)].copy()
# Проверяем полученные результаты
df_actual.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,wii sports,Wii,2006.0,sports,41.36,28.96,3.77,8.45,76.0,8.0,E
2,mario kart wii,Wii,2008.0,racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,wii sports resort,Wii,2009.0,sports,15.61,10.93,3.28,2.95,80.0,8.0,E
6,new super mario bros.,DS,2006.0,platform,11.28,9.14,6.50,2.88,89.0,8.5,E
7,wii play,Wii,2006.0,misc,13.96,9.18,2.93,2.84,58.0,6.6,E


### Вывод к отфильтрованным данным 

Для дальнейшего анализа был выбран период с 2000 по 2013 года включительно, так как именно этот период интересует заказчиков исследования.

Отфильтрованные данные сохранены в отдельный датафрейм `df_actual`.

---

## Категоризация данных
   

### Категоризация игр в зависимости от оценок пользователя <a class="tocSkip">

In [26]:
def user_score_category(score):
    if score >= 8:
        return 'высокая оценка'
    elif score >= 3:
        return 'средняя оценка'
    else:
        return 'низкая оценка'
# Добавляем новый столбец
df_actual['user_score_category'] = df_actual['user_score'].apply(user_score_category)

### Категоризация игр в зависимости от оценок критиков <a class="tocSkip">

In [27]:
def critic_score_category(score):
    if score >= 80:
        return 'высокая оценка'
    elif score >= 30:
        return 'средняя оценка'
    else:
        return 'низкая оценка'
# Добавляем новый столбец
df_actual['critic_score_category'] = df_actual['critic_score'].apply(critic_score_category)

### Группировка и подсчет игр по выделенным категориям оценки <a class="tocSkip">

In [28]:
# Группируем и считаем количество игр по категориям оценок пользователей 
df_actual.groupby('user_score_category')['name'].count()

user_score_category
высокая оценка    2286
низкая оценка     6414
средняя оценка    4081
Name: name, dtype: int64

In [29]:
# Группируем и считаем количество игр по категориям оценок критиков 
df_actual.groupby('critic_score_category')['name'].count()

critic_score_category
высокая оценка    1692
низкая оценка     5667
средняя оценка    5422
Name: name, dtype: int64

### Топ-7 платформ по количеству выпущенных игр за актуальный период<a class="tocSkip">

In [30]:
top_platforms = df_actual['platform'].value_counts().head(7)

top_platforms

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: count, dtype: int64

### Результаты по категоризации игр 

После фильтрации данных за период 2000–2013 годов была проведена категоризация игр по оценкам пользователей и критиков.

Наибольшее количество игр относится к категории с низкими пользовательскими оценками. Это может говорить о том, что в рассматриваемый период выпускалось большое количество игр, которые не получали высоких оценок пользователей.

По оценкам критиков наблюдается преобладание игр со средними и низкими оценками. Количество игр с высокими оценками критиков значительно меньше, что подтверждает тот факт, что действительно высокооценённые игры встречаются реже.

Также был проведён анализ количества игр по платформам. Самой популярной платформой по количеству выпущенных игр оказалась PS2. 


## Результаты исследования

В рамках проекта были подготовлены данные для анализа рынка видеоигр:

- проведена очистка исходного датасета;
- исправлены типы данных;
- обработаны пропуски и дубликаты;
- сформирован аналитический срез за период 2000–2013 годов;
- созданы категории игр по оценкам пользователей и критиков;
- определены наиболее распространенные игровые платформы.

---

## Итоговый вывод


В рамках исследования был проведён анализ рынка видеоигр. 
Исходный датасет содержал 16956 строк и 11 столбцов с информацией о названии игры, платформе, годе выпуска, жанре, 
продажах в различных регионах, а также оценках пользователей и критиков и возрастном рейтинге.

На первом этапе была выполнена загрузка данных и первичное знакомство с ними. 
Анализ структуры датасета показал наличие пропущенных значений, некорректных типов данных и неунифицированных названий столбцов.

В процессе предобработки данных были выполнены следующие действия:
- названия столбцов приведены к стилю snake_case для удобства работы;
- некоторые столбцы с числовыми данными (eu_sales, jp_sales, user_score) были преобразованы из типа object в числовой тип;
- строковые значения в числовых столбцах были заменены на пропуски с помощью параметра errors='coerce';
- были проанализированы пропущенные значения и принято решение удалить строки с пропусками в столбцах name и genre, 
так как они не имеют аналитической ценности;
- проведена нормализация текстовых данных: жанры и названия игр приведены к нижнему регистру, 
а значения рейтинга — к верхнему регистру;
- выявлены и удалены явные дубликаты строк.

В результате предобработки было удалено 243 строки, что составляет примерно 1.43% от исходного объёма данных.

Далее был сформирован аналитический срез данных. 
Для исследования был выбран период с 2000 по 2013 год включительно, так как именно этот временной 
промежуток интересует заказчиков исследования. Отфильтрованные данные были сохранены в отдельный датафрейм df_actual.

На следующем этапе была проведена категоризация игр по оценкам пользователей и критиков. 
В датафрейм были добавлены два новых столбца:
- user_score_category — категория оценки пользователей (высокая, средняя, низкая);
- critic_score_category — категория оценки критиков (высокая, средняя, низкая).

Анализ распределения игр по категориям показал, что наибольшее количество игр относится к категории 
с низкими пользовательскими оценками. Это может свидетельствовать о том, что в рассматриваемый период 
выпускалось большое количество игр среднего и невысокого качества.

Аналогичная тенденция наблюдается и в оценках критиков: большинство игр имеет средние и низкие оценки, 
тогда как игры с высокими оценками встречаются значительно реже.

Также был проведён анализ количества игр, выпущенных на различных платформах. 
Были определены топ-7 платформ по количеству выпущенных игр в период с 2000 по 2013 год. 
Лидером по количеству игр оказалась платформа PS2, на которой было выпущено 2127 игр. 
Далее следуют DS и Wii. 

Таким образом, проведённый анализ позволил подготовить структурированные данные и получить общее представление 
о развитии игровой индустрии в рассматриваемый период. Результаты исследования могут быть использованы при 
подготовке аналитической статьи о тенденциях развития игровой индустрии и популярности игровых платформ в начале XXI века.